### DISCUSSION TITLE
Highly concurrent seems make log's tid repeated

### DISCUSSION THREAD

**Discussion by ilovedumplings** [apache/skywalking#9918] [Q&A]
my english are no good.
my project use logback-spring.xml to set log config .
based   on  net.logstash.logback.appender.LogstashTcpSocketAppender  and use org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider  to provide TID 

when i see the kibana's log . on seconds same TID show so many times above 500 times . this definitely not once request. 
i'm confused is about skywalking or logstash's issues. (probably logstash , beacause TCP may cause lose packet , but it doesn't seem likely to make TID repeated )

this things occur to others??
somebody can give me some advice?

**Comment by wu-sheng** [comment]
I could just say SkyWalking doesn't have duplicated trace ID issue from design perspective.
But your tracing context could be leaked due to incompatible plugins or wrong usage.
This is a thing could be answered as yes or no, I am afraid. This is not how things work.
You could try to find whether this trace ID(repeated one) makes sense from trace view. If not, then you know that is one or botg of things I mentioned above.

**Reply by ilovedumplings** [reply]
我今天在本地复现了一下,线程并发数设置的是60,然后TID分别在控制台输出和logststash输出,控制台输出的TID是正常的,不包含任何多余链路,但是输出到logstash之后,就出现了TID出现在多个线程的情况。

--------------------------------------------------
on my own machine , use apipost test concurrent , is's happened every time , on the idea's console everything is ok (TID only showed on one thread ) , use elk , TID showed in different  threads.

here is my logback-spring.xml 



<appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
            <layout class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.mdc.TraceIdMDCPatternLogbackLayout">
                <Pattern>%d{HH:mm:ss.SSS} [%thread] %-5level [%X{tid}] %logger - %msg%n</Pattern>
            </layout>
        </encoder>
    </appender>
    
    <appender name="logstash"
              class="net.logstash.logback.appender.LogstashTcpSocketAppender">
        <writeBufferSize>1048576</writeBufferSize>
        <destination>${LOGSTASH_DESTINATION}</destination>
        <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
            <providers class="net.logstash.logback.composite.loggingevent.LoggingEventJsonProviders">
                <timestamp>
                    <timeZone>UTC</timeZone>
                </timestamp>
                <pattern>
                    <pattern>
                        {
                        "severity": "%level",
                        "thread": "%thread",
                        "class": "%logger{40}",
                        "rest": "%message %ex",
                        "service": "${SPRING_APPLICATION_NAME:-}",
                        "exception": "%ex",
                        }
                    </pattern>
                </pattern>
                <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
                    <fieldName>TID</fieldName>
                </provider>
        </encoder>
    </appender>

这是我appender基础的配置

**Reply by ilovedumplings** [reply]
```
<appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
            <layout class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.mdc.TraceIdMDCPatternLogbackLayout">
                <Pattern>%d{HH:mm:ss.SSS} [%thread] %-5level [%X{tid}] %logger - %msg%n</Pattern>
            </layout>
        </encoder>
    </appender>
    <!-- logstash appender-->
    <appender name="logstash"
              class="net.logstash.logback.appender.LogstashTcpSocketAppender">
        <writeBufferSize>1048576</writeBufferSize>
        <destination>${LOGSTASH_DESTINATION}</destination>
        <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
            <providers class="net.logstash.logback.composite.loggingevent.LoggingEventJsonProviders">
            <timestamp>
                <timeZone>UTC</timeZone>
            </timestamp>
            <pattern>
                <pattern>
                    {
                    "severity": "%level",
                    "thread": "%thread",
                    "class": "%logger{40}",
                    "rest": "%message %ex",
                    "service": "${SPRING_APPLICATION_NAME:-}",
                    "exception": "%ex",
                    }
                </pattern>
            </pattern>
            <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
                <fieldName>TID</fieldName>
            </provider>
        </encoder>
    </appender>

**Reply by wu-sheng** [reply]
Sorry, we don't have any experience in this kind of use case. Log frameworks could include a lot of async mechanisms, which may or may not mix up the context.
So, I am afraid you have to do debugging on your own. Only you have this scenario, and you have your private env. 
We could only help if you have found the issue, and provide suggestions based on that.
